### PR TITLE
implement TP for group convolution

### PR DESCRIPTION
**Pull request by justheuristic**
This pull request implements tensor parallelism for ConvNd and ConvTransposeNd with groups != 1

### CODE REVIEW

File: src/tensor_parallel/factory.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from typing import Optional, Sequence
 
 import torch
+import torch.distributed
```
Comment (justheuristic) [inline src/tensor_parallel/factory.py:5]: why: torch.distributed may not be loaded in some cases. if it isnt, the local usage of torch.distributed would fail

File: tests/test_basic.py [hunk 2]
```diff
@@ -48,17 +48,16 @@ def test_convs(devices, extra_options):
...
         model_tp = TensorParallel(model_tp, device_ids=devices)
-        out_ours = model_tp(inputs)
+        out_ours = model_tp(inputs2)
         out_ours.norm().backward()
-        assert torch.allclose(ref_out, out_ours, atol=1e-6)
-        dim = 1 if model[0].transposed else 0  # concat over output channels
-        our_grad = torch.cat([next(shard[0].parameters()).grad for shard in model_tp.module_shards], dim=dim)
-        assert torch.allclose(model[0].weight.grad, our_grad, atol=1e-6)
+        assert torch.allclose(ref_out, out_ours, atol=1e-6), abs(ref_out - out_ours).max()
+        assert torch.allclose(inputs1.grad, inputs2.grad, atol=1e-5), abs(inputs1.grad - inputs2.grad).max()
```
Comment (justheuristic) [inline tests/test_basic.py:63]: __why was this test changed?:__ for grouped convolutions, the weight tensor is sharded in a less-than-straightforward way. Hence, it is no longer sufficient to concatenate gradients to obtain the full gradient: you need to shuffle individual grads around. Instead, this code checks that grads are correct w.r.t. an input vector whose shape does not depend on layer type.

File: src/tensor_parallel/slicer_wrapper.py [hunk 3]
```diff
@@ -88,16 +89,33 @@ def get_default_config(cls, module: nn.Module) -> Config:
...
-                config.state_rules[f"^{name}.bias$"] = "split 0"
+                if module.bias is not None:
+                    config.state_rules[f"^{name}.bias$"] = "split 0"
                 config.output_rules[f"^{name}$"] = {0: "gather 1"}
             elif isinstance(module, conv._ConvNd) and module.groups != 1:
-                logger.warning(
-                    f"AutoConfig does not support sharding convolution layers {name} with {module.groups} groups yet; "
-                    "If you're sure that you need it, implement sharding in place of this message and open a PR"
-                )
+                # group conv: split each group individually over input channels to avoid changing module.groups
```
Comment (justheuristic) [inline src/tensor_parallel/slicer_wrapper.py:96]: Why is it important to not change num_groups? if num_groups % world_size != 0, ranks would get an uneven number of groups which would require a custom convolution operation. If user wants this behavior for their specific model, we humbly recommend them to write their own config :)

**Review by BlackSamorez — APPROVED** [review]

**Review by IaroslavLisniak — APPROVED** [review]
